### PR TITLE
Creating Chatbot Using Natural Language Processing in Python

### LINKED ISSUE #5932: Creating ChatBot Using Natural Language Processing in Python
### Topic Suggestion
Creating ChatBot Using Natural Language Processing in Python
#### Proposed article introduction
Bots are pieces of custom-written software that engage in conversation with other internet users on their behalf. Algorithms are the building blocks of bots, which aid them in their work. This means they are self-contained and can begin the process of conservation without the need for human intervention.

We will be designing a chatbot using natural language processing.

#### Key takeaways 
- Type of Bots
- Building the chatbot
- Benefits of bots

#### Article quality
In this article, we'll learn about Creating ChatBot Using Natural Language Processing in Python. The code will be easy to read and understand.
#### References
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by vincentkims49**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5932

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by vincentkims49** [comment]
@BrianaNzivu Here is the PR

**Comment by louisefindlay23** [comment]
Plagiarism check complete

File: content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
+A bot is a computer program that performs predetermined tasks automatically. It is the goal of the bots to perform human duties in the same way that humans did in the past. In a nutshell, they mimic human behaviour.
+
+We'll design a conversational interface for our chatbot using natural language processing in this tutorial.
+
+### Prerequisites
+- Have a basic knowledge of [Python](https://www.python.org/).
+- Have python environment of your choice installed.
```
Comment (BrianaNzivu) [inline content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md]: IDE?

File: content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+- **Spam Bots:** Designed to gather email addresses from spam mailing lists.
+- **Spider Bots:** Searches the web and downloads pages before indexing the material on each page it comes across.
+- **Transnational Bots:** Designed to be used in transactions.
+- **Social Media Bot:** Automatically responds to questions on social media platforms.
+
+### Building an NLP chatbot
+The first step is to create an intents JSON file that lists all the possible outcomes of user interactions with our chatbot. We first need a set of tags that users can use to categorize their queries. These tags include `name`, `age` etc. Every new tag would require its own unique pattern. Identifying these trends can help the chatbot train itself on how people query about our chatbot's name, allowing it to be more responsive. To answer questions, the chatbot will return pre-programmed responses.
+
+#### Importing modules
+Let's import the necessary modules first:
```
Comment (BrianaNzivu) [inline content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md]: Which are?
List them down

File: content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+- **Scraper Bots:** They are designed to read and save data from the Internet.
+- **Spam Bots:** Designed to gather email addresses from spam mailing lists.
+- **Spider Bots:** Searches the web and downloads pages before indexing the material on each page it comes across.
+- **Transnational Bots:** Designed to be used in transactions.
+- **Social Media Bot:** Automatically responds to questions on social media platforms.
+
+### Building an NLP chatbot
+The first step is to create an intents JSON file that lists all the possible outcomes of user interactions with our chatbot. We first need a set of tags that users can use to categorize their queries. These tags include `name`, `age` etc. Every new tag would require its own unique pattern. Identifying these trends can help the chatbot train itself on how people query about our chatbot's name, allowing it to be more responsive. To answer questions, the chatbot will return pre-programmed responses.
+
+#### Importing modules
```
Comment (BrianaNzivu) [inline content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md:55]: Clearly highlight your steps. I assume this is step one. Do the same for the rest.

File: content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+import tensorflow as tensorF # A multidimensional array of elements is represented by this symbol.
+from tensorflow.keras import Sequential # Sequential groups a linear stack of layers into a tf.keras.Model
+from tensorflow.keras.layers import Dense, Dropout
+
+nltk.download("punkt")# required package for tokenization
+nltk.download("wordnet")# word database
+```
+
+#### Creating jason file
+Next we will create the json file:
```
Comment (BrianaNzivu) [inline content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md:74]: A JSON file that will do what?

**Review by BrianaNzivu — CHANGES_REQUESTED** [review]
Hey @vincentkims49 👋 

First of all, good job on the article. It is a very interesting read.

Just a few changes I need you to make that I have mentioned as we continue with the preliminary revisions.
1.  Please structure your article according to this [sample](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md). I need you to clearly highlight your steps and explain them well.
2.  Refer to the single comments I have made.

**Note: Please make sure to make changes to the updated index.md**

In case of any question or help needed please do not hesitate to ping me here :smile:.

Also, ping me when done.

Regards.

**Comment by vincentkims49** [comment]
@BrianaNzivu Done

File: content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+- **Spider Bots:** Searches the web and downloads pages before indexing the material on each page it comes across.
+- **Transnational Bots:** Designed to be used in transactions.
+- **Social Media Bot:** Automatically responds to questions on social media platforms.
+
+### Building an NLP chatbot
+Follow the steps below to successfully build a conversational interface for our chatbot.
+
+#### Step 1: Importing modules
+Imports are critical for successfully organising your Python code. Importing code correctly will increase your productivity by allowing you to reuse code while also maintaining the maintainability of your projects.
+
```
Comment (BrianaNzivu) [inline content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md:52]: I wanted you to list down the modules you are importing.

Comment (masese911) [reply content/articles/creating-chatbot-using-natural-language-processing-in-python/index.md:52]: Okay I'll list them right now

**Comment by vincentkims49** [comment]
@BrianaNzivu please have a look

**Comment by BrianaNzivu** [comment]
Proceeding to tech review 👍

**Comment by BrianaNzivu** [comment]
Tech review done 👍

**Review by BrianaNzivu — APPROVED** [review]
Hey @vincentkims49 👋 .

Very good job on the article. It is a very interesting read👏 .

Just a few things I'd like to point out:
- I suggest that you check out the [structure](https://github.com/section-engineering-education/engineering-education/tree/master/new_contributors) for the various articles accepted by Section.
- Lastly please go through ALL the corrections I have made so that you can not repeat the same mistakes.

That's it! :smile: 

A final review will be done now by a content moderator.
Otherwise, congratulations on publishing your article🍾.

In case of any question or clarification (currently or in the future), do not hesitate to reach out via [Slack](https://app.slack.com/client/T5X14SH7G/D01A6MT673R) or my personal email (briananzivu@gmail.com). 
I am **ALWAYS** here to help.

It has been a pleasure reading and reviewing this article.

Looking forward to seeing you contribute more to EngEd. 🚀

Once again good job! 👍 

Kind Regards

**Pull request merged**